src/compiler/scala/tools/nsc/backend/jvm/analysis/NullnessAnalyzer.scala

```diff
@@ -0,0 +1,282 @@
+case object Unknown extends Nullness
+case object Null    extends Nullness
+
+/**
+ * Represents the nullness state for a local variable or stack value.
+ *
+ * Note that nullness of primitive values is not tracked, it will be always [[Unknown]].
+ */
+sealed trait NullnessValue extends Value {
+  /**
+   * The nullness of this value.
+   */
+  def nullness: Nullness
+
+  /**
+   * True if this value is a long or double. The Analyzer framework needs to know
+   * the size of each value when interpreting instructions, see `Frame.execute`.
+   */
+  def isSize2: Boolean
+  /**
+   * The size of the slot described by this value. Cannot be 0 because no values are allocated
+   * for void-typed slots, see NullnessInterpreter.newValue.
+   **/
+  def getSize: Int = if (isSize2) 2 else 1
+
+  def merge(other: NullnessValue) = NullnessValue(nullness merge other.nullness, isSize2)
+}
+
+object NullValue     extends NullnessValue { def nullness = Null;    def isSize2 = false; override def toString = "Null"     }
+object UnknownValue1 extends NullnessValue { def nullness = Unknown; def isSize2 = false; override def toString = "Unknown1" }
+object UnknownValue2 extends NullnessValue { def nullness = Unknown; def isSize2 = true;  override def toString = "Unknown2" }
+object NotNullValue  extends NullnessValue { def nullness = NotNull; def isSize2 = false; override def toString = "NotNull"  }
+
+object NullnessValue {
+  def apply(nullness: Nullness, isSize2: Boolean): NullnessValue = {
+    if      (nullness == Null)    NullValue
+    else if (nullness == NotNull) NotNullValue
+    else if (isSize2)             UnknownValue2
+    else                          UnknownValue1
+  }
+
+  def apply(nullness: Nullness, insn: AbstractInsnNode): NullnessValue = {
+    apply(nullness, isSize2 = BytecodeUtils.instructionResultSize(insn) == 2)
+  }
+}
+
+final class NullnessInterpreter extends Interpreter[NullnessValue](Opcodes.ASM5) {
+  def newValue(tp: Type): NullnessValue = {
+    // ASM loves giving semantics to null. The behavior here is the same as in SourceInterpreter,
+    // which is provided by the framework.
+    //
+    // (1) For the void type, the ASM framework expects newValue to return `null`.
+    //     Also, the Frame.returnValue field is `null` for methods with return type void.
+    //     Example callsite passing VOID_TYPE: in Analyzer, `newValue(Type.getReturnType(m.desc))`.
+    //
+    // (2) `tp` may also be `null`. When creating the initial frame, the analyzer invokes
+    //     `newValue(null)` for each local variable. We have to return a value of size 1.
+    if (tp == Type.VOID_TYPE) null // (1)
+    else NullnessValue(Unknown, isSize2 = tp != null /*(2)*/ && tp.getSize == 2 )
+  }
+
+  override def newParameterValue(isInstanceMethod: Boolean, local: Int, tp: Type): NullnessValue = {
+    // For instance methods, the `this` parameter is known to be not null.
+    if (isInstanceMethod && local == 0) NullnessValue(NotNull, isSize2 = false)
```

Comment:
Thanks, that's a good point. The code here is correct because only sets the initial value of the local variable `0`. If it is re-assigned within a method, its nullness will be updated in the frame after the store instruction.